// @lwc/template-compiler/dist/commonjs/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
const errors_1 = require("@lwc/errors");
const shared_1 = require("@lwc/shared");
const DEFAULT_CONFIG = {
    secure: false,
    experimentalDynamicDirective: false,
    experimentalComputedMemberExpression: false,
    format: 'module',
};
const AVAILABLE_OPTION_NAMES = new Set([
    'secure',
    'experimentalComputedMemberExpression',
    'experimentalDynamicDirective',
    'stylesheetConfig',
]);
function mergeConfig(config) {
    errors_1.invariant(config !== undefined && typeof config === 'object', errors_1.TemplateErrors.OPTIONS_MUST_BE_OBJECT);
    for (const property in config) {
        if (!AVAILABLE_OPTION_NAMES.has(property) && shared_1.hasOwnProperty.call(config, property)) {
            throw errors_1.generateCompilerError(errors_1.TemplateErrors.UNKNOWN_OPTION_PROPERTY, {
                messageArgs: [property],
            });
        }
    }
    return {
        ...DEFAULT_CONFIG,
        ...config,
    };
}
exports.mergeConfig = mergeConfig;
